Update the analyze databases article (#10459)
github/codeql: docs/codeql/codeql-cli/analyzing-databases-with-the-codeql-cli.rst

Before starting an analysis you must:

- :doc:`Set up the CodeQL CLI <getting-started-with-the-codeql-cli>` to run commands locally.
- :doc:`Create a CodeQL database <creating-codeql-databases>` for the source
  code you want to analyze.

> It is simplest to run ``database analyze`` using CodeQL packs. You can also

Reviewer: Should this suggestion be(?):

```suggestion
The simplest way to run ``codeql database analyze`` is using CodeQL packs. You can also
```

> It is simplest to run ``database analyze`` using CodeQL packs. You can also

Author: Yes!

> It is simplest to run ``database analyze`` using CodeQL packs. You can also

Reviewer: Cool, applied!

> It is simplest to run ``database analyze`` using CodeQL packs. You can also

Reviewer: This is discussed in another thread/suggestion so I'll go ahead and resolve this.

Before starting an analysis you must:

- :doc:`Set up the CodeQL CLI <getting-started-with-the-codeql-cli>` to run commands locally and

>   optionally check out the CodeQL repository if you want direct access to the CodeQL core queries

Reviewer 2: Avoid highlighting the idea of checking out the repo here. If they are using a bundle or packs there is no need.

Author: Good point.

You can analyze a database by running the following command::

>    codeql database analyze <database> --format=<format> --output=<output> <query-specifiers>...

Reviewer: That's fine, thanks for the context! Feel free to resolve this comment with no change.

You can analyze a database by running the following command::

>    codeql database analyze <database> --format=<format> --output=<output> ...<query-specifiers>

Author: The `...` indicates that there can be multiple query specifiers. There should be some visual indicator that multiple specifiers can be used, but maybe `...` isn't best.

>    codeql database analyze <database> --format=<format> --output=<output> ...<query-specifiers>

Author: The docs for [database analyze](https://codeql.github.com/docs/codeql-cli/manual/database-analyze/) uses `...` _after_ the parameter. I'll do the same.

```suggestion
   codeql database analyze <database> --format=<format> --output=<output> <query-specifiers>...
```

You can also specify:

> - ``...<query-specifiers>``: a list of queries to run over your database. This

Author: I'll follow what we do for database analyze. I also want to add a few words.

```suggestion
- ``<query-specifiers>...``: a spece-separated list of queries to run over your database. This
```

> - ``...<query-specifiers>``: a list of queries to run over your database. This

Reviewer 3: 👍🏻 

Note there's a typo: spece-separated -> space-separated

> - ``...<query-specifiers>``: a list of queries to run over your database. This

Author: 😢

You can also specify:

> - ``<query-specifiers>...``: a spece-separated list of queries to run over your database. This

Author: ```suggestion
- ``<query-specifiers>...``: a space-separated list of queries to run over your database. This
```

For full details of all the options you can use when analyzing databases, see
the `database analyze reference documentation <../manual/database-analyze>`__.

.. _specifying-which-queries:

Specifying which queries to run in a CodeQL pack
------------------------------------------------

Query specifiers are used by ``codeql database analyze`` and other commands that operate on a set of queries.

> The complete way to specify a set of queries is in the form ``scope/name@range:path``, where:

Reviewer: ```suggestion
The complete form of a query specifier is``scope/name@range:path``, where:
```

Minor, removes some superfluous language.

- ``scope/name`` is the qualified name of a CodeQL pack.
- ``range`` is a `semver range <https://docs.npmjs.com/cli/v6/using-npm/semver#ranges>`_.
- ``path`` is a file system path to a single query, a directory containing queries, or a query suite file.

When you specify a ``scope/name``, the ``range`` and ``path`` are
optional. If you omit a ``range`` then the latest version of the
specified pack is used. If you omit a ``path`` then the default query suite
of the specified pack is used.

> The ``path`` can be one of a ``.ql`` query file, a directory

Reviewer: ```suggestion
The ``path`` can be one of: a ``.ql`` query file, a directory
```

V minor.

containing one or more queries, or a ``.qls`` query suite file. If
you omit a pack name, then you must provide a ``path``,
which will be interpreted relative to the working directory
of the current process. Glob patterns are not supported.

> If you specify a ``scope/name`` and ``path``, then the ``path`` cannot
> be absolute. It is considered relative to the root of the CodeQL
> pack.

Reviewer: ```suggestion
If you specify both a ``scope/name`` and ``path``, then the ``path`` cannot
be absolute. It is considered relative to the root of the CodeQL
pack.
```

V minor.

* ``codeql/python-queries:Functions`` - All queries in the ``Functions`` directory in the latest version of the ``codeql/python-queries`` pack.
* ``codeql/python-queries@1.2.3:Functions`` - All queries in the ``Functions`` directory in version 1.2.3 of the ``codeql/python-queries`` pack.
* ``codeql/python-queries@1.2.3:codeql-suites/python-code-scanning.qls`` - All queries in the ``codeql-suites/python-code-scanning.qls`` directory in version 1.2.3 of the ``codeql/python-queries`` pack.
* ``suites/my-suite.qls`` - All queries in the ``suites/my-suite.qls`` file relative to the current working directory.

.. pull-quote::

    Tip

    The default query suite of the standard CodeQL query packs are ``codeql-suites/<lang>-code-scanning.qls``. Several other useful query suites can also be found in the ``codeql-suites`` directory of each pack. For example, the ``codeql/cpp-queries`` pack contains the following query suites:

    * ``cpp-code-scanning.qls`` - Standard Code Scanning queries for C++. The default query suite for this pack.

>     * ``cpp-security-extended.qls`` - Security-extended queries for C++. This suite contains queries that are less precise than the standard security queries, and may find more false-positives. This query suite includes all queries from ``cpp-code-scanning.qls``.
>     * ``cpp-security-and-quality.qls`` - Security-and-quality queries for C++. This query suite includes all queries from ``cpp-security-extended.qls``.

Reviewer 2: Adapted the descriptions from https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#running-additional-queries.
```suggestion
    * ``cpp-security-extended.qls`` - Queries from the default  ``cpp-code-scanning.qls`` suite for C++, plus lower severity and precision queries.
    * ``cpp-security-and-quality.qls`` - Queries from ``cpp-security-extended.qls``, plus maintainability and reliability queries.
```

    You can see the sources for these query suites in the `CodeQL repository <https://github.com/github/codeql/tree/main/cpp/ql/src/codeql-suites>`__. Query suites for other languages are similar.

> Examples of using ``codeql database analyze``

Author: How about this:

```suggestion
Examples of running database analyses
```

> Examples of using ``codeql database analyze``

Reviewer: LGTM

---------------------------------------------

The following examples show how to run ``database analyze`` using CodeQL packs as
well as using a local checkout of the CodeQL repository. These examples assume your
CodeQL databases have been created in a directory that is a sibling of your local
copies of the CodeQL repository.

.. _run-query-pack:

Running a CodeQL query pack
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. include:: ../reusables/beta-note-package-management.rst

To run an existing CodeQL query pack from the GitHub Container registry, you can specify one or more
pack names::

   codeql database analyze <database> microsoft/coding-standards@1.0.0 github/security-queries --format=sarifv2.1.0 --output=query-results.sarif --download

The ``analyze`` command above runs the default query suite of two CodeQL query packs: ``microsoft/coding-standards`` version 1.0.0 and the latest version of ``github/security-queries`` on the
specified database. For further information about default suites, see

> ":ref:`Publishing and using CodeQL packs <publishing-and-using-codeql-packs>`".

Reviewer: Will resolve this as it's newer style and as discussed, this doc follows an older one.

Query specifiers are used by ``codeql database analyze`` and other commands that operate on a set of queries.
The complete way to specify a set of queries is in the form ``scope/name@range:path``, where:

- ``scope/name`` is the qualified name of a CodeQL pack.
- ``range`` is a `semver range <https://docs.npmjs.com/cli/v6/using-npm/semver#ranges>`_.
- ``path`` is a file system path to a single query, a directory containing queries, or a query suite file.

When you specify a ``scope/name``, the ``range`` and ``path`` are
optional. If you omit a ``range`` then the latest version of the
specified pack is used. If you omit a ``path`` then the default query suite
of the specified pack is used.

The ``path`` can be one of a ``.ql`` query file, a directory
containing one or more queries, or a ``.qls`` query suite file. If
you omit a pack name, then you must provide a ``path``,
which will be interpreted relative to the working directory
of the current process. Glob patterns are not supported.

If you specify a ``scope/name`` and ``path``, then the ``path`` cannot
be absolute. It is considered relative to the root of the CodeQL
pack.

Example query specifiers
~~~~~~~~~~~~~~~~~~~~~~~~

> * ``codeql/python-queries`` - All the queries in the default query suite of the latest version of the ``codeql/python-queries`` pack.

Reviewer 2: Should we mention that the default suite is `python-code-scanning.qls`?

.. _database-analyze-examples:

.. _specifying-which-queries:

Specifying which queries to run in a CodeQL pack
------------------------------------------------

> *Query specifiers* are used by ``codeql database analyze`` and other commands that operate on a set of queries.

Reviewer 3: ```suggestion
Query specifiers are used by ``codeql database analyze`` and other commands that operate on a set of queries.
```

For more information, see `Analyzing a CodeQL database <https://docs.github.com/en/code-security/code-scanning/using-codeql-code-scanning-with-your-existing-ci-system/configuring-codeql-cli-in-your-ci-system#analyzing-a-codeql-database>`__
or `Code scanning API <https://docs.github.com/en/rest/reference/code-scanning>`__ in the GitHub documentation.

CodeQL query suites are ``.qls`` files that use directives to select queries to run
based on certain metadata properties. The standard CodeQL packs have metadata that specify
the location of the query suites used by code scanning, so the CodeQL CLI knows where to find these
suite files automatically, and you don't have to specify the full path on the command line.
For more information, see ":ref:`Creating CodeQL query suites <creating-codeql-query-suites>`."

Query suites in the core CodeQL query packs are all stored in the ``codeql-suites`` directory of the pack. For example, the ``codeql/cpp-queries`` pack contains the following query suites:

* ``cpp-code-scanning.qls`` - Standard Code Scanning queries for C++.
* ``cpp-security-and-quality`` - Security-and-quality queries for C++.

> * ``cpp-security-extended`` - Security-extended queries for C++. This suite contains queries that are less precise than the standard security queries, and may find more false-positives.

Reviewer 2: Reorder these: security-extended should go first, and highlight that each includes the suites above it.
